File: tools/features/failure-analysis/actions/verify_reproduction.action.ts
import type {
  FailureAnalysisRequest,
  FailureInvestigationContext,
  FailureTerminalState,
} from "@tools-contracts/failure-analysis";

import type { FailureAnalysisResponse } from "../models/failure_analysis.model";
import { redactFailureFingerprint } from "../shared/redact_failure_display";
import { postFailureSidecar } from "../shared/sidecar_failure_client";

type VerifyReproductionInput = Extract<
  FailureAnalysisRequest,
  { action: "verify_reproduction" }
>["input"];

export async function verifyReproductionAction(
  input: VerifyReproductionInput,
): Promise<FailureAnalysisResponse> {
  if ("terminalState" in input) {
    return terminalResponse(input.terminalState, input.investigation);
  }
  try {
    const response = await postFailureSidecar(
      input.sidecarBaseUrl,
      "__probe/failure/verify",
      {
        captureId: input.captureId,
        expectedExceptionType: input.expectedFingerprint.exceptionType,
        expectedRootCauseType: input.expectedFingerprint.rootCauseType,
        expectedNearestApplicationMethodKey: input.expectedFingerprint.nearestApplicationMethodKey,
      },
      input.timeoutMs,
      input.sidecarAuthorization,
    );
    if (response.status < 200 || response.status >= 300 || response.json === null) {
      if (response.status === 404) return captureUnavailableResponse(input.captureId);
      return unavailableResponse(response.status);
    }
    const matched = response.json.outcome === "matched";
    const structuredContent = {
      outcome: matched ? "REPRODUCED" : "NOT_REPRODUCED",
      reasonCode: matched ? "ok" : String(response.json.outcome ?? "inconclusive"),
      expectedFingerprint: input.expectedFingerprint,
      observedFingerprint: redactFailureFingerprint(response.json.observedFingerprint),
      lineHit: input.lineHit,
      attemptEvidence: { captureId: input.captureId, sidecarOutcome: response.json.outcome },
      cleanupStatus: "external_workflow_owned",
      ...(input.investigation ? { investigation: input.investigation } : {}),
      ...(matched ? {} : { diagnosisClaimed: false }),
    };
    const text = matched
      ? "Failure reproduced with matching runtime fingerprint and Strict Line Key evidence."
      : "Failure was not reproduced. No diagnosis is claimed.";
    return { content: [{ type: "text", text }], structuredContent };
  } catch {
    return unavailableResponse(undefined);
  }
}

function terminalResponse(
  terminalState: FailureTerminalState,
  investigation: FailureInvestigationContext | undefined,
): FailureAnalysisResponse {
  const structuredContent = {
    outcome: terminalState.outcome,
    reasonCode: terminalState.reasonCode,
    attemptEvidence: { attemptCount: terminalState.attemptCount },
    cleanupStatus: terminalState.cleanupStatus,
    diagnosisClaimed: false,
    ...(investigation ? { investigation } : {}),
  };
  return {
    content: [
      {
        type: "text",
        text: "Failure investigation ended without runtime reproduction. No diagnosis is claimed.",
      },
    ],
    structuredContent,
  };
}

function captureUnavailableResponse(captureId: string): FailureAnalysisResponse {
  const structuredContent = {
    outcome: "INCONCLUSIVE",
    reasonCode: "capture_not_found",
    attemptEvidence: { captureId },
    diagnosisClaimed: false,
  };
  return {
    content: [
      {
        type: "text",
        text: "Failure verification is inconclusive because the capture is unavailable.",
      },
    ],
    structuredContent,
  };
}

function unavailableResponse(status: number | undefined): FailureAnalysisResponse {
  const structuredContent = {
    outcome: "BLOCKED_SIDECAR_UNAVAILABLE",
    reasonCode: "sidecar_failure_verification_unavailable",
    ...(typeof status === "number" ? { httpStatus: status } : {}),
  };
  return {
    content: [
      { type: "text", text: "Failure verification is blocked because the Sidecar is unavailable." },
    ],
    structuredContent,
  };
}

File: mcp-server/core/src/main/java/com/nimbly/mcpjavadevtools/server/core/feature/failureanalysis/action/impl/VerifyReproductionAction.java
package com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.action.impl;

import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.action.FailureAnalysisActionHandler;
import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.endpoint.FailureEvidenceClient;
import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.endpoint.FailureEvidenceClientException;
import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.endpoint.FailureEvidenceFailureKind;
import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.endpoint.FailureEvidenceResponseMapper;
import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.fingerprint.FailureFingerprintComparator;
import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.model.action.FailureAnalysisAction;
import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.model.action.verifyreproduction.VerifyReproductionRequest;
import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.model.endpoint.FailureEvidenceResponse;
import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.model.endpoint.FailureVerificationEvidence;
import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.model.endpoint.FailureVerifyEvidenceRequest;
import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.model.request.FailureAnalysisRequest;
import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.model.result.FailureAnalysisOutcome;
import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.model.result.FailureAnalysisResult;
import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.model.result.FailureAttemptEvidence;
import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.model.result.FailureCleanupStatus;
import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.model.result.FailureVerificationDetails;
import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.model.terminal.FailureTerminalState;
import com.nimbly.mcpjavadevtools.server.core.feature.failureanalysis.policy.FailureAnalysisPolicy;

/** Real runtime and terminal-state verify_reproduction action. */
public final class VerifyReproductionAction implements FailureAnalysisActionHandler {

    private final FailureEvidenceClient client;
    private final FailureEvidenceResponseMapper responseMapper;
    private final FailureAnalysisPolicy policy;

    public VerifyReproductionAction(
            FailureEvidenceClient client,
            FailureEvidenceResponseMapper responseMapper,
            FailureAnalysisPolicy policy) {
        this.client = client;
        this.responseMapper = responseMapper;
        this.policy = policy;
    }

    @Override
    public FailureAnalysisAction action() {
        return FailureAnalysisAction.VERIFY_REPRODUCTION;
    }

    @Override
    public FailureAnalysisResult execute(FailureAnalysisRequest input) {
        if (!(input instanceof VerifyReproductionRequest request)) {
            return FailureAnalysisResult.invalidRequest();
        }
        if (request.terminalState() != null) {
            return terminal(request);
        }
        return runtime(request);
    }

    private FailureAnalysisResult runtime(VerifyReproductionRequest request) {
        try {
            FailureEvidenceResponse response = client.verify(new FailureVerifyEvidenceRequest(
                    request.sidecarBaseUrl(), request.captureId(), request.expectedFingerprint().exceptionType(),
                    request.expectedFingerprint().rootCauseType(),
                    request.expectedFingerprint().nearestApplicationMethodKey(),
                    request.sidecarAuthorization(), policy.timeoutOrDefault(request.timeout())));
            return result(response, request);
        } catch (FailureEvidenceClientException exception) {
            if (exception.failureKind() == FailureEvidenceFailureKind.INTERRUPTED) {
                Thread.currentThread().interrupt();
            }
            return FailureAnalysisResult.blockedVerification(null, request.investigation());
        } catch (RuntimeException exception) {
            return FailureAnalysisResult.blockedVerification(null, request.investigation());
        }
    }

    private FailureAnalysisResult result(
            FailureEvidenceResponse response, VerifyReproductionRequest request) {
        if (response.status() == 404) {
            return FailureAnalysisResult.captureNotFound(request.captureId(), request.investigation());
        }
        if (response.status() < 200 || response.status() >= 300 || response.payload() == null) {
            return FailureAnalysisResult.blockedVerification(response.status(), request.investigation());
        }
        FailureVerificationEvidence evidence = responseMapper.verify(response.payload());
        if (evidence.outcome() == null) {
            return FailureAnalysisResult.invalidVerification(request.investigation());
        }
        FailureAttemptEvidence attempts = new FailureAttemptEvidence(
                request.captureId(), evidence.outcome(), null);
        String mismatch = FailureFingerprintComparator.mismatchReason(
                request.expectedFingerprint(), evidence.observedFingerprint());
        if ("matched".equals(evidence.outcome()) && mismatch == null) {
            return FailureAnalysisResult.verification(
                    FailureAnalysisOutcome.REPRODUCED, "ok", new FailureVerificationDetails(
                            request.expectedFingerprint(), evidence.observedFingerprint(), request.lineHit(),
                            attempts, request.investigation(), true));
        }
        String reason = "matched".equals(evidence.outcome()) && mismatch != null
                ? mismatch : evidence.outcome();
        return FailureAnalysisResult.verification(
                FailureAnalysisOutcome.NOT_REPRODUCED, reason, new FailureVerificationDetails(
                        request.expectedFingerprint(), evidence.observedFingerprint(), request.lineHit(),
                        attempts, request.investigation(), false));
    }

    private FailureAnalysisResult terminal(VerifyReproductionRequest request) {
        FailureTerminalState state = request.terminalState();
        return FailureAnalysisResult.terminal(
                terminalOutcome(state.outcome()), state.reasonCode(),
                new FailureAttemptEvidence(null, null, state.attemptCount()),
                cleanupStatus(state.cleanupStatus()), request.investigation());
    }

    private static FailureAnalysisOutcome terminalOutcome(String value) {
        for (FailureAnalysisOutcome outcome : FailureAnalysisOutcome.values()) {
            if (outcome.value().equals(value)) {
                return outcome;
            }
        }
        throw new IllegalArgumentException("unsupported terminal outcome");
    }

    private static FailureCleanupStatus cleanupStatus(String value) {
        return switch (value) {
            case "cleanup_confirmed" -> FailureCleanupStatus.CLEANUP_CONFIRMED;
            case "cleanup_incomplete" -> FailureCleanupStatus.CLEANUP_INCOMPLETE;
            case "external_workflow_owned" -> FailureCleanupStatus.EXTERNAL_WORKFLOW_OWNED;
            default -> throw new IllegalArgumentException("unsupported cleanup status");
        };
    }
}
